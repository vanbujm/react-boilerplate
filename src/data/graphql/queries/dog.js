/* eslint-disable no-plusplus */
import { capitalize } from 'lodash';
import DOG_BREEDS from '../../../constants/dogs/dogBreeds';
import dogNames from '../../../constants/dogs/dogNames';

const seededRandom = (max, min, seed) => {
  const actualMax = max || 1;
  const actualMin = min || 0;

  // some prime numbers to make things exciting
  const actualSeed = (seed * 9301 + 49297) % 233280;
  const rnd = actualSeed / 233280;

  return actualMin + rnd * (actualMax - actualMin);
};

const createDogs = (numberOfDoggos = 20, seed = 2) => {
  const doggos = [];
  let newSeed = seed;

  for (let i = 0; i < numberOfDoggos; i++) {
    const breedIndex = Math.floor(seededRandom(DOG_BREEDS.length, 0, newSeed));
    const nameIndex = Math.floor(
      seededRandom(DOG_BREEDS.length, 0, newSeed + 1),
    );
    const doggo = {
      id: i + 1,
      name: capitalize(dogNames[nameIndex]),
      breed: DOG_BREEDS[breedIndex],
      isGoodDog: 'Yes',
    };
    doggos.push(doggo);
    newSeed++;
  }
  return doggos;
};

const doggos = createDogs();

const dog = (_, args, _2, { cacheControl }) => {
  const oneMinute = 60;

  cacheControl.setCacheHint({ maxAge: oneMinute });
  const id = Number(args.id);
  if (!id) {
    return null;
  }

  const doggo = doggos.find(aDog => aDog.id === id);

  return new Promise(resolve => {
    setTimeout(() => resolve(doggo !== undefined ? doggo : null), 3000);
  });
};

export { dog, doggos };
