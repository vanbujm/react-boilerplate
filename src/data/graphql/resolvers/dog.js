/* eslint-disable no-plusplus */
import dogs from '../../../constants/dogs';

const dog = (_, args) => {
  const id = Number(args.id);
  if (!id) {
    return null;
  }

  const doggo = dogs.find(aDog => aDog.id === id);

  return new Promise(resolve => {
    setTimeout(() => resolve(doggo !== undefined ? doggo : null), 3000);
  });
};

export default dog;
