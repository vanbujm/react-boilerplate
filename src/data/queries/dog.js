/* eslint-disable no-plusplus */
import faker from 'faker';
import { GraphQLID as ID } from 'graphql';
import DogType from '../types/DogType';
import DOG_BREEDS from '../../constants/dogBreeds';

const doggos = [];

const numberOfDoggos = 20;

for (let i = 0; i < numberOfDoggos; i++) {
  const doggo = {
    id: i + 1,
    name: faker.name.firstName(),
    breed: DOG_BREEDS[Math.floor(Math.random() * numberOfDoggos)],
    isGoodDog: true,
  };
  doggos.push(doggo);
}

const dog = {
  type: DogType,
  args: {
    id: {
      type: ID,
    },
  },
  resolve(_, args, _2, { cacheControl }) {
    const oneMinute = 60;

    cacheControl.setCacheHint({ maxAge: oneMinute });
    const id = Number(args.id);
    if (!id) {
      return null;
    }

    const doggo = doggos.find(aDog => aDog.id === id);

    return new Promise(resolve => {
      setTimeout(() => resolve(doggo !== undefined ? doggo : null), 3000);
    });
  },
};

export { dog, doggos };
