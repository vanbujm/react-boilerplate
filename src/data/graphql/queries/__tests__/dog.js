/* eslint-env jest */
import 'babel-polyfill';
import { dog } from '../dog';

const setCacheHintSpy = jest.fn();

describe('test dog query', () => {
  test('matches expected query object', async () => {
    const dog2 = await dog(null, { id: 2 }, null, {
      cacheControl: { setCacheHint: setCacheHintSpy },
    });
    await expect(dog2).toMatchSnapshot();
  });
});
