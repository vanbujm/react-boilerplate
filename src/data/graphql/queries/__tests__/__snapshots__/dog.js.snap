// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test dog query matches expected query object 1`] = `
Object {
  "breed": "Dutch Shepherd Dog",
  "id": 2,
  "isGoodDog": "Yes",
  "name": "Hunter",
}
`;
